### PR TITLE
Feature/uar 1142 get bo trust links

### PR DESCRIPTION
**Pull request by azafarCH**
### JIRA link

https://companieshouse.atlassian.net/browse/UAR-1142

### Change description

Retrieving (hashed) BO trust links data from private-sdk-java.

### Work checklist

- [ ] Tests added where applicable

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/PrivateDataRetrievalService.java [hunk 1]
```diff
@@ -216,7 +223,37 @@ public PrivateTrustDetailsListApi getTrustDetails(String companyNumber) throws S
           ApiLogger.errorContext(message, e);
           throw new ServiceException(e.getMessage(), e);
         }
-      }
+    }
+
+    public PrivateTrustLinksListApi getTrustLinks(String companyNumber) throws ServiceException {
+        var logMap = new HashMap<String, Object>();
+        logMap.put(COMPANY_NUMBER, companyNumber);
+        ApiLogger.info("Retrieving BO Trust Links for Company Number " + companyNumber, logMap);
```
Comment (RobertBromfield) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/PrivateDataRetrievalService.java:231]: Change "BO" to "Beneficial Owners"

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/PrivateDataRetrievalService.java [hunk 2]
```diff
@@ -216,7 +223,37 @@ public PrivateTrustDetailsListApi getTrustDetails(String companyNumber) throws S
...
+        try {
+        PrivateTrustLinksListApi trustsLinks = apiClientService.getInternalApiClient()
+                .privateTrustLinksResourceHandler()
+                .getTrustLinks(OVERSEAS_ENTITY_URI_SECTION + companyNumber + "/trusts/beneficial-owners/links")
+                .execute()
+                .getData();
+
+        if (trustsLinks != null && trustsLinks.getData() != null && !trustsLinks.getData().isEmpty()) {
+            ApiLogger.info(String.format("Retrieved %d BO Trust links for Company Number %s",
```
Comment (RobertBromfield) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/PrivateDataRetrievalService.java:240]: Change "BO" to "Beneficial Owners"

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/PrivateDataRetrievalService.java [hunk 3]
```diff
@@ -216,7 +223,37 @@ public PrivateTrustDetailsListApi getTrustDetails(String companyNumber) throws S
...
+                    trustsLinks.getData().size(), companyNumber));
+        }
+
+        return trustsLinks;
+        } catch (ApiErrorResponseException e) {
+        if (e.getStatusCode() == HttpServletResponse.SC_NOT_FOUND) {
+            ApiLogger.info("No BO Trust links found for Company Number " + companyNumber, logMap);
```
Comment (RobertBromfield) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/PrivateDataRetrievalService.java:247]: Change "BO" to "Beneficial Owners"

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/PrivateDataRetrievalService.java [hunk 4]
```diff
@@ -216,7 +223,37 @@ public PrivateTrustDetailsListApi getTrustDetails(String companyNumber) throws S
...
+            return new PrivateTrustLinksListApi(Collections.emptyList());
+        }
+        throw new ServiceException(e.getStatusMessage(), e);
+        } catch (URIValidationException e) {
+          var message = "Error Retrieving BO Trust links data for " + companyNumber;
```
Comment (RobertBromfield) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/PrivateDataRetrievalService.java:252]: Change "BO" to "Beneficial Owners"

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/service/PrivateDataRetrievalServiceTest.java [hunk 5]
```diff
@@ -584,6 +615,18 @@ void testGetTrustDetailsApiErrorResponseExceptionThrownNotFoundReturnsEmptyList(
             assertEquals(0, result.getData().size());
         }
 
+        @Test
+        void testGetTrustLinksApiErrorResponseExceptionThrownNotFoundReturnsEmptyList() throws ApiErrorResponseException, URIValidationException, ServiceException {
+            var exception = new ApiErrorResponseException(
+                    new HttpResponseException.Builder(404, "notFound", new HttpHeaders()));
```
Comment (RobertBromfield) [inline src/test/java/uk/gov/companieshouse/overseasentitiesapi/service/PrivateDataRetrievalServiceTest.java:621]: Here you can use the constant `FOUR_HUNDRED_AND_FOUR_EXCEPTION` 

Looking at it there is actually two one incorrectly named `FourHundredAndFourException` if you could rename that one `FOUR_HUNDRED_AND_FOUR_EXCEPTION` then delete the original that would be great

**Review by RobertBromfield — CHANGES_REQUESTED** [review]

**Review by RobertBromfield — APPROVED** [review]

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/configuration/InterceptorConfig.java [hunk 6]
```diff
@@ -27,14 +27,16 @@ public class InterceptorConfig implements WebMvcConfigurer {
     static final String BENEFICIAL_OWNERS_PRIVATE_DATA = "/private/**/beneficial-owners";
     static final String TRUST_DETAILS_PRIVATE_DATA = "/private/**/trusts/details";
     static final String CORPORATE_TRUSTEES_PRIVATE_DATA = "/private/**/corporate-trustees";
+    static final String TRUST_LINKS_PRIVATE_DATA = "/private/**/trusts/beneficial-owner/links";
```
Comment (bbell-companieshouse) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/configuration/InterceptorConfig.java:30]: Maybe could have this as just 
```suggestion
    static final String TRUST_LINKS_PRIVATE_DATA = "/private/**/links";
```

**Review by bbell-companieshouse — APPROVED** [review]

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/configuration/InterceptorConfig.java [hunk 7]
```diff
@@ -27,14 +27,16 @@ public class InterceptorConfig implements WebMvcConfigurer {
     static final String BENEFICIAL_OWNERS_PRIVATE_DATA = "/private/**/beneficial-owners";
     static final String TRUST_DETAILS_PRIVATE_DATA = "/private/**/trusts/details";
     static final String CORPORATE_TRUSTEES_PRIVATE_DATA = "/private/**/corporate-trustees";
+    static final String TRUST_LINKS_PRIVATE_DATA = "/private/**/links";
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/configuration/InterceptorConfig.java:30]: static final String TRUST_LINKS_PRIVATE_DATA = "/private/**/trusts/beneficial-owners/links";

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/configuration/InterceptorConfig.java (hunk 6 above)
Comment (AndreKramer7654) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/configuration/InterceptorConfig.java]: Would be better to be explicit about what we can match though.

**Pull request merged**